### ISSUE TITLE
`thread_local!` initialization code panics on some `aarch64-apple-darwin` envs, works on others

### ISSUE THREAD

**Issue by dj8yfo** [rust-lang/rust#132704]
I tried this code:

https://github.com/near/near-sdk-rs/blob/master/near-sdk/src/environment/mock/mod.rs#L11-L16

```rust
thread_local! {
    static BLOCKCHAIN_INTERFACE: RefCell<MockedBlockchain>
         = RefCell::new(MockedBlockchain::default());
}
```

I expected to see this happen: 

`thread_local!` initialization code for `RefCell::new(MockedBlockchain::new)`
works

Instead, this happened: 

on unclarified `aarch64-apple-darwin` environment(s) `thread_local!`
panics on https://doc.rust-lang.org/src/core/ptr/mod.rs.html#1277 .

The issue doesn't reproduce on other `aarch64-apple-darwin` real macs, 
and on `macos-14-arm64` github actions vm.

Original issue: https://github.com/near/near-sdk-rs/issues/1252 

### Meta

`rustc --version --verbose`:

```
stable: 1.80, 1.81, 1.82 panics
beta 1.83 panics
nightly 1.84 panics
```

backtrace from original issue: 
https://github.com/near/near-sdk-rs/issues/1252#issuecomment-2454692564

**Comment by dj8yfo** [comment]
### Somewhat limited pre-analysis

`ub_checks::assert_unsafe_precondition!` check inside of
https://github.com/rust-lang/rust/blob/master/library/std/src/sys/thread_local/native/lazy.rs#L66 
fails, 
because earlier `LazyStorage::new` failed to put `State::Initial` 
into an adress with correct alignment for  `State<T,D>` 
https://github.com/rust-lang/rust/blob/master/library/std/src/sys/thread_local/native/lazy.rs#L37,
where `T` is `RefCell<MockedBlockchain<Memory = MockedMemory>>` 


### Somewhat related case

earlier we had a similar issue with `aarch64-apple-darwin` with somewhat similar panics:

```bash
0x102f569ec - core::ptr::replace::he9e2c42f9fe96d9b
                  at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/ub_checks.rs:73:17
0x102f569ec - core::ptr::mut_ptr::<impl *mut T>::replace::hcc3425bab485c184
                  at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/ptr/mut_ptr.rs:1560:18
0x102f569ec - std::sys::thread_local::lazy::LazyKeyInner<T>::take::h11414c1b64d26b2a
                  at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/std/src/sys/thread_local/mod.rs:102:30
```

which originated in some `register_dtor` functions when trying to run doc-tests
for the same `near-sdk-rs` project : https://github.com/rust-lang/rust/issues/126643 

It was later confirmed that the issue stopped to manifest itself on exactly the same
`rustc 1.79.0 (129f3b996 2024-06-10)` compiler after `macos-14-arm64` github actions
vm was updated.

**Comment by madsmtm** [comment]
I tried going through the linked issues, but I couldn't figure out how you reproduce the issue (what commands do you run?).

A hypothetical minimal reproducer based on the code you posted doesn't reproduce on my Mac (M2, macOS 14.7):
```rust
use std::cell::RefCell;

#[repr(align(16))]
#[derive(Debug)]
struct MockedBlockchain {
    _size: [u8; 2512],
}

thread_local! {
    static BLOCKCHAIN_INTERFACE: RefCell<MockedBlockchain> = RefCell::new(MockedBlockchain {
        _size: [1; 2512],
    });
}

fn main() {
    BLOCKCHAIN_INTERFACE.with(|b| println!("{b:?}"));
}
```

So it sounds like it might be something else that is causing the issue?

**Comment by madsmtm** [comment]
> on **unclarified** `aarch64-apple-darwin` environment(s) `thread_local!` panics on https://doc.rust-lang.org/src/core/ptr/mod.rs.html#1277 .

What do you mean by "unclarified"? Knowing that might help us to reproduce the issue?

**Comment by dj8yfo** [comment]
@madsmtm you might ask @peitalin from the linked issue about specifics of his environment, where he reproduced it. We didn't reproduce it on our macs either. 

The command sequence to run the same scenario which triggered panic would be

```bash
# install cargo-near to $CARGO_HOME/bin
curl --proto '=https' --tlsv1.2 -LsSf https://github.com/near/cargo-near/releases/latest/download/cargo-near-installer.sh | sh
cargo near new test
cd test/
cargo test
```

**Comment by dj8yfo** [comment]
@madsmtm the code you tried is much simpler than the one from `near-sdk` repo. One from `near-sdk` repo uses generic parameter and also some `unsafe` to create references to self (without ouroborous). The self-referential part has been there for some odd 3 years, while the generic parameter has been added recently, a couple of months ago.

**Comment by madsmtm** [comment]
@peitalin: Could you try with `-Clinker=/usr/bin/ld` (just to make sure that the problem isn't with Clang installed via Homebrew or similar)?

Also, what's your OS version (output of `sw_vers`)?

**Comment by madsmtm** [comment]
Linking possibly related https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/Weird.20crash.20using.20thread.20locals.20Apple.20Silicon.20debug.

So I'll also ask:
- What version of Xcode are you using?
- What version of `ld` are you using?

**Comment by RalfJung** [comment]
Rust recently gained extra assertions in debug builds to detect Undefined Behavior caused by misaligned pointers. That is likely why this started failing. Already on older versions of Rust, this code had Undefined Behavior, and could be the cause of subtle misbehavior -- newer Rust just makes this misbehavior easier to detect.

**Comment by madsmtm** [comment]
It takes quite a lot of time to test these things, Xcode is large ;)

But lemme write down exactly what I've tested so far:
- WORKS: Command Line Tools for Xcode 14.3.1
- UNKNOWN: Xcode 14.3.1
- BROKEN: Command Line Tools for Xcode 15.0
- UNKNOWN: Xcode 15.0
- UNKNOWN: Xcode 15.0.1
- BROKEN: Command Line Tools for Xcode 15.1
- UNKNOWN: Xcode 15.1
- UNKNOWN: Xcode 15.2
- WORKS: Command Line Tools for Xcode 15.3
- UNKNOWN: Xcode 15.3
- WORKS: Xcode 15.4

And in between these there's a lot of betas and release candidates.

> EDIT: Actually https://developer.apple.com/documentation/xcode-release-notes/xcode-15_2-release-notes says that XCode 15.2 should work on macOS 13.5.

Yeah, I should've clarified that I only really tested the Command Line Tools (and no Command Line Tools for Xcode 15.2 exist).

**Comment by madsmtm** [comment]
> Is this issue documented/tracked somewhere public?

Not that I know of? Maybe rdar://110340167, though I don't have access to that myself, just guessing based on [a similar issue found by googling](https://forums.developer.apple.com/forums/thread/731090)?

> Are there newer linker versions available?

With a newer Xcode, yes. Without, no (not practically, at least).

> I don't know anything about macOS, I am just surprised that an OS released just one year ago wouldn't be able to run the latest version of the linker.

Xcode generally only wants to be run on the latest macOS release. Usually it works fine in older versions, but you have to edit a configuration file and run an "unsupported" build, so most people don't.

Also linking possibly related https://github.com/rust-lang/rust/issues/90959.

**Comment by madsmtm** [comment]
You're probably right that this is / has been UB for a long time. A bisection led me to a6cdd81eff52566542cecdc1ce381dbe42cf77fb, which sounds very much like a red herring (the only thing of relevance that PR changed is a bit of panic code, which might explain why alignment would change between that and the previous commit).

I'll try to take a look at the difference in assembly between a version that succeeds and one that fails, but it's difficult, since it seems like it requires the standard library to be linked in a certain way (I couldn't trigger the issue with `-Zbuild-std`).

**Comment by saethlin** [comment]
Yeah this looks exactly like what I previously poked at on Zulip, it's somewhat comforting that this seems to depend on the Apple Xcode bag of tools.

> which sounds very much like a red herring

FWIW, the MIR inliner used to use item hash comparisons to prevent query cycles, which meant that literally any edit to a crate would cause the inliner to make different decisions. The only thing you can do in this scenario is turn off the MIR inliner with `RUSTFLAGS=-Zinline-mir=no` or `RUSTFLAGS=-Zmir-enable-passes=-Inline`. But of course the standard library in the range you're trying to bisect through has been precompiled with the unstable MIR optimization pipeline.

So you'd need to make `-Zbuild-std` work to get a usable bisection. By default, it builds the standard library with the optimization settings for the rest of the build. So if you don't set `--release`, you get a quite different standard library build. Are you just using `cargo run -Zbuild-std` without `--release`?

**Comment by madsmtm** [comment]
> FWIW, the MIR inliner used to use item hash comparisons to prevent query cycles, which meant that literally any edit to a crate would cause the inliner to make different decisions. The only thing you can do in this scenario is turn off the MIR inliner with `RUSTFLAGS=-Zinline-mir=no` or `RUSTFLAGS=-Zmir-enable-passes=-Inline`. But of course the standard library in the range you're trying to bisect through has been precompiled with the unstable MIR optimization pipeline.

That explains a lot, thanks for the tips!

> Are you just using `cargo run -Zbuild-std` without `--release`?

Yup, couldn't reproduce this issue at all with `--release`, but perhaps that's just 'cause I didn't try hard enough.
